### ISSUE TITLE
Rework caching of blobs in the registry

### ISSUE THREAD

**Issue by miminar** [openshift/origin#10217]
There are two caches with very similar purpose:
### Upstream Inmemory cache

Docker registry is, by default, deployed with `inmemory` blobdescriptor cache. This cache stores permanently associations of:

```
<localrepository>/<blobdigest>    <->    distribution.Descriptor{Digest string; Size int64; MediaType string}
```

The entries are stored until an API endpoint `DELETE /v2/<name>/blobs/<digest>` is called or registry is restarted. Unfortunately, a lot can happen on etcd side without ever hitting the registry API, so the cache can get quickly out of date. For example:
1. image is untagged from image stream
2. image stream is removed
3. image is removed

None of the above will remove the corresponding associations from the `inmemory` cache. So the user will be able to stat the affected blobs easily.
### Digest cache

Is a shared cache of blob digests to remote and local repositories that have previously been identified as containing that blob. It contains two
kinds of associations:
1. `<blobdigest> <-> <registry>/<namespace>/<name>`
2. `<blobdigest> <-> <namespace>/<name>`

The first associates a blob with a remote repository. Such an entry is set and used by pullthrough middleware. The second associates a blob with a local repository. Such a blob is expected to reside on local storage. It's set and used by blobDescriptorService middleware.

It resides in registry middleware and is fully under our control.

As opposed to `inmemory` cache, this one evicts its entries after configurable
amount of time.
### The Problem

Apart from limitations of upstream inmemory cache, there's another problem. Both caches serve very similar purpose which means the entries are duplicated which leads to excessive memory usage.
### Possible solution

We should use just one cache. Ideally, we should be able to implement cache.BlobDescriptorCacheProvider interface and register that with upstream registry application. Upstream currently does not support registering own blobdescriptor cache providers but that could be easily changed with a small upstream patch.

Resulting cache should inherit the best from both:
1. evict stale entries
2. distinguish between local and remote repositories
3. contain whole blob descriptor
4. apply both to layer links and blobs in global blob store
5. be easily accessible so we can add new entries as we see fit
6. be configurable from registry's config file

**Comment by legionus** [comment]
> It contains two kinds of associations:
>
>   {blobdigest} = {registry}/{namespace}/{name}
>   {blobdigest} = {namespace}/{name}

But the second index is first actually. I mean if we know our address (`{registry}`) we do not need second one. Our address in `DOCKER_REGISTRY_URL` environment variable and for origin it in `OPENSHIFT_DEFAULT_REGISTRY`.

**Comment by miminar** [comment]
> But the second index is first actually. I mean if we know our address ({registry}) we do not need second one. Our address in DOCKER_REGISTRY_URL environment variable and for origin it in OPENSHIFT_DEFAULT_REGISTRY.

It happens that registry gets re-deployed. Since `DOCKER_REGISTRY_URL` usually refers to internal IP that can be changed during the re-deploy, I don't find it reliable to depend on this prefix.

I remember @smarterclayton wanted to establish some special docker image referece prefix that would allow us to easily identify managed image just by its reference. I'd be in favor of such a change.

**Comment by legionus** [comment]
> I remember @smarterclayton wanted to establish some special docker image referece prefix that would allow us to easily identify managed image just by its reference. I'd be in favor of such a change.

But according https://github.com/openshift/origin/issues/8581#issuecomment-239880317 we can use `docker-registry.default.svc:5000`.

**Comment by miminar** [comment]
> But according #8581 (comment) we can use docker-registry.default.svc:5000.

I don't think we can assume that registry is deployed in `default` namespace, running at 5000 port or that the service is named `docker-registry`.

**Comment by legionus** [comment]
Port number stored in `DOCKER_REGISTRY_SERVICE_PORT`, namespace can be only `default` (https://github.com/openshift/origin/issues/2581) and name will be `docker-registry` (https://github.com/openshift/origin/issues/8849)

**Comment by liggitt** [comment]
that is not the long term plan, don't add more things that depend on those assumptions

**Comment by smarterclayton** [comment]
You can't use the DNS name yet.  Either way it needs to be injected to you,
not specified inside or assumed inside your code.

**Comment by legionus** [comment]
> You can't use the DNS name yet. Either way it needs to be injected to you, not specified inside or assumed inside your code.

@smarterclayton @miminar We can use predefined constant (`localhost` for exmaple) to indicate that blobdigest from local registry. Inside docker-regisrty this name does not matter. The origin may substitute this constant with the DNS name or IP address then use it.

**Comment by smarterclayton** [comment]
For the cache may be ok.  But that value can't change without a restart of the registry, so a cache would be flushed.

**Comment by liggitt** [comment]
we also should be careful about assumptions in the registry that the registry is a singleton

**Comment by dmage** [comment]
Closed by openshift/image-registry#35.
